lib/Dialect/AIEX/IR/AIEXDialect.cpp

```diff
@@ -110,6 +173,34 @@ LogicalResult AIEX::NpuDmaMemcpyNdOp::verify() {
     return emitOpError("Stride 2 exceeds the [1:1M] range.");
   if (strides[0] > 0x100000)
     return emitOpError("Stride 1 exceeds the [1:1M] range.");
+
+  if (offset % 4 != 0) {
+    return emitOpError("Offset must be 4-byte-aligned.");
+  }
+
+  bool error = false;
+  std::stringstream msg;
+  for (int i = 0; i < 3; i++) {
+    if (raw_strides[i] * elemWidth % addressGranularity != 0) {
+      error = true;
+      msg << "Stride " << i << " is " << raw_strides[i] << " elements * "
+          << (elemWidth / 8) << " bytes = " << (raw_strides[i] * elemWidth / 8)
+          << " bytes, which is not divisible by 4. ";
```

Comment:
Is this 4 because of the address granularity? Should it be a different number if the address granularity is more than 4 bytes?

Reply:
You're right, I should use the addressGranularity variable for the error message here. All hardware currently uses 4, but that doesn't mean it shouldn't be changeable.